### PR TITLE
Resolve #773 -- Root buff added

### LINKED ISSUE #773: Buffs doesn't work
ObjAIBase.cs in:
```
public void ApplyCrowdControl(UnitCrowdControl cc)
        {
            if (cc.IsTypeOf(CrowdControlType.ROOT) || cc.IsTypeOf(CrowdControlType.STUN))
            {
                StopMovement(); //  this line should delete
                //Should be: 
                //Stats.SetActionState(ActionState.CAN_MOVE,false);
                //Stats.SetActionState(ActionState.CAN_NOT_MOVE, true);
            }
            _crowdControlList.Add(cc);
        }
 public void RemoveCrowdControl(UnitCrowdControl cc)
        {
            /*should be:
            if (cc.IsTypeOf(CrowdControlType.ROOT) ||  || cc.IsTypeOf(CrowdControlType.STUN))
            {
                
                Stats.SetActionState(ActionState.CAN_MOVE, true);
                Stats.SetActionState(ActionState.CAN_NOT_MOVE, false);
            }*/
                _crowdControlList.Remove(cc);
        }
```

And should be AttacableUnit.cs:
 ```
public void StopMovement()
        {
            SetWaypoints(new List<Vector2> { GetPosition(), GetPosition() });
        }

        public override void Update(float diff)
        {
            /*should be:
            if (!Stats.GetActionState(ActionState.CAN_MOVE) || Stats.GetActionState(ActionState.CAN_NOT_MOVE))
            {
                StopMovement();
            }
            else
            {    
               base.Update(diff);
            }*/
            _statUpdateTimer += diff;
            while (_statUpdateTimer >= 500)
            {
                // update Stats (hpregen, manaregen) every 0.5 seconds
                Stats.Update(_statUpdateTimer);
                _statUpdateTimer -= 500;
            }
        }
```

### PR DESCRIPTION
**Pull request by requizm**
Normally buffs(root or all other) doesn't work. I added root.
These changes require the root buff to work.

ref https://github.com/LeagueSandbox/GameServer/issues/773

### CODE REVIEW

**Comment by cajthamlmatej** [comment]
Please read CONTRIBUTING.md before contributing, create issue first.

**Comment by Veykril** [comment]
Is there a reason as to why you removed the stun check and replaced it by only checking for roots?

**Comment by requizm** [comment]
> Is there a reason as to why you removed the stun check and replaced it by only checking for roots?

Actually, i don't have a definite suggestion. Maybe we could write a separate function to each crowd control

```
if(cc == Root)
{
rootfunc();
}
if(cc == Stun)
{
stunfunc();
}
```

But I know that crowd control system has to change from the beginning. Almost nothing has been written.

**Comment by Veykril** [comment]
ye, the cc system and the buff system in general probably needs a complete
overhaul

File: GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs [hunk 1]
```diff
@@ -230,17 +230,24 @@ public void RemoveBuffSlot(Buff b)
 
         public void ApplyCrowdControl(UnitCrowdControl cc)
         {
-            if (cc.IsTypeOf(CrowdControlType.STUN) || cc.IsTypeOf(CrowdControlType.ROOT))
+            if (cc.IsTypeOf(CrowdControlType.ROOT))
             {
-                StopMovement();
+                Stats.SetActionState(ActionState.CAN_MOVE,false);
```
Comment (MythicManiac) [inline GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs]: How about creating a function
```CS
public void LockMovement()
{
    Stats.SetActionState(ActionState.CAN_MOVE, false);
    Stats.SetActionState(ActionState.CAN_NOT_MOVE, true);
    StopMovement();
}
```
And calling it here?

Also add back the stun check, or an additional if check for the stun that does the same thing, however you see fit better.

File: GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs [hunk 2]
```diff
@@ -230,17 +230,24 @@ public void RemoveBuffSlot(Buff b)
...
 
             _crowdControlList.Add(cc);
         }
 
         public void RemoveCrowdControl(UnitCrowdControl cc)
         {
-            _crowdControlList.Remove(cc);
+            if (cc.IsTypeOf(CrowdControlType.ROOT))
+            {
+                Stats.SetActionState(ActionState.CAN_MOVE, true);
```
Comment (MythicManiac) [inline GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs]: Likewise, a function here:
```CS
public void UnlockMovement()
{
    Stats.SetActionState(ActionState.CAN_MOVE, true);
    Stats.SetActionState(ActionState.CAN_NOT_MOVE, false);
}
```

**Comment by danil179** [comment]
Im not sure if this is should be part of IObjBase or AttackableUnit. Change your function to 2 functions as suggested to avoid confusion. It seems that your function disabling the calls to base class which can be problematic, and also calling every tick to stop movement which is not needed (just create variable CanMove or something similar).

**Comment by danil179** [comment]
Seems like #1036 implemented most of these buffs types using `CanMove` function. I'm currently closing this PR and its associated issue #773.

**Pull request closed without merging**